### PR TITLE
More mock testing

### PR DESCRIPTION
**Pull request by justinsb**
Create some golden tests to try to make sure our apply library is reasonably efficient / so we can track what it is doing.

- tests: Improve mock kubeapiserver, introduce test for direct Applier
- tests: record and validate the requests we are making
- tests: add test for when objects already exist
- tests: add test for the case when last-applied is present

### CODE REVIEW

**Comment by justinsb** [comment]
Most of the lines of code are from https://github.com/kubernetes/kops/tree/master/pkg/applylib/mocks/mockkubeapiserver ...  I'm trying "copy and paste" reuse until we figure out where it should live.  But maybe it should just live here, so maybe I should give it its own go.mod here...

File: pkg/mocks/mockkubeapiserver/getresource.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	if err != nil {
+		return err
+	}
+	if !found {
+		return req.writeErrorResponse(http.StatusNotFound)
+	}
+
+	j, err := json.Marshal(object)
+	if err != nil {
+		klog.Warningf("object does not marshal: %v", err)
```
Comment (cheftako) [inline pkg/mocks/mockkubeapiserver/getresource.go:63]: Is there a reason we aren't using structured logging?

File: pkg/mocks/mockkubeapiserver/putresource.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	resourceRequestBase
+}
+
+// Run serves the http request
+func (req *putResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
+	gr := schema.GroupResource{Group: req.Group, Resource: req.Resource}
+
+	id := types.NamespacedName{Namespace: req.Namespace, Name: req.Name}
+
+	existing, found, err := s.storage.GetObject(ctx, gr, id)
```
Comment (cheftako) [inline pkg/mocks/mockkubeapiserver/putresource.go:42]: existing makes me think its a bool. oldVal? existingVal?

File: pkg/patterns/declarative/pkg/applier/direct.go [hunk 3]
```diff
@@ -46,8 +46,14 @@ func (d *directApplier) NewBuilder(opt ApplierOptions) *resource.Builder {
 	return resource.NewBuilder(restClientGetter)
 }
 
-func (d *directApplier) NewFactory() cmdutil.Factory {
-	return cmdutil.NewFactory(&genericclioptions.ConfigFlags{})
+func (d *directApplier) NewFactory(opt ApplierOptions) cmdutil.Factory {
+	var configFlags genericclioptions.ConfigFlags
+	if opt.RESTConfig != nil {
```
Comment (cheftako) [inline pkg/patterns/declarative/pkg/applier/direct.go:51]: Might be worth a comment on why this is important.
I believe this is preventing unnecessary repeated openapi calls when a restconfig exists and has already made those calls.

**Comment by cheftako** [comment]
/lgtm
Couple of minor questions but on the whole looks good.

File: pkg/mocks/mockkubeapiserver/putresource.go (hunk 2 above)
Comment (justinsb) [reply pkg/mocks/mockkubeapiserver/putresource.go:42]: Will change to existingVal - thank you!

File: pkg/mocks/mockkubeapiserver/getresource.go (hunk 1 above)
Comment (justinsb) [reply pkg/mocks/mockkubeapiserver/getresource.go:63]: Blocked on logr not supporting Warning: https://github.com/go-logr/logr/issues/151

**Pull request merged**